src/binder_radio.c

```diff
@@ -0,0 +1,553 @@
+    } else {
+        binder_radio_check_state(self);
+    }
+}
+
+static
+void
+binder_radio_power_request_cb(
+    RadioRequest* req,
+    RADIO_TX_STATUS status,
+    RADIO_RESP resp,
+    RADIO_ERROR error,
+    const GBinderReader* args,
+    void* user_data)
+{
+    BinderRadioObject* self = THIS(user_data);
+
+    GASSERT(self->pending_req == req);
+    radio_request_unref(self->pending_req);
+    self->pending_req = NULL;
+
+    if (status == RADIO_TX_STATUS_OK) {
+        if (resp != RADIO_RESP_SET_RADIO_POWER) {
+            ofono_error("Unexpected setRadioPower response %d", resp);
+        } else if (error != RADIO_ERROR_NONE) {
+            ofono_error("Power request failed: %s",
+                binder_radio_error_string(error));
+        }
+    } else {
+        ofono_error("Power request failed");
+    }
+
+    binder_radio_power_request_done(self);
+}
+
+static
+void
+binder_radio_submit_power_request(
+    BinderRadioObject* self,
+    gboolean on)
+{
+    /* setRadioPower(int32 serial, bool on) */
+    GBinderWriter writer;
+    RadioRequest* req = radio_request_new(self->client,
+        RADIO_REQ_SET_RADIO_POWER, &writer,
+        binder_radio_power_request_cb, NULL, self);
+
+    gbinder_writer_append_bool(&writer, on);
+
+    self->next_state_valid = FALSE;
+    self->next_state = on;
+    self->state_changed_while_request_pending = 0;
+    binder_radio_cancel_retry(self);
+
+    GASSERT(!self->pending_req);
+    radio_request_set_blocking(req, TRUE);
+    if (radio_request_submit(req)) {
+        self->pending_req = req; /* Keep the ref */
+    } else {
+        radio_request_unref(req);
+    }
+}
+
+static
+void
+binder_radio_power_request(
+    BinderRadioObject* self,
+    gboolean on,
+    gboolean allow_repeat)
+{
+    const char* on_off = on ? "on" : "off";
+
+    if (self->pending_req) {
+        if (allow_repeat || self->next_state != on) {
+            /* Wait for the pending request to complete */
+            self->next_state_valid = TRUE;
+            self->next_state = on;
+            DBG_(self, "%s (queued)", on_off);
+        } else {
+            DBG_(self, "%s (ignored)", on_off);
+        }
+    } else {
+        if (binder_radio_state_on(self->last_known_state) == on) {
+            DBG_(self, "%s (already)", on_off);
+            binder_radio_check_state(self);
+        } else {
+            DBG_(self, "%s", on_off);
+            binder_radio_submit_power_request(self, on);
+        }
+    }
+}
+
+static
+void
+binder_radio_state_changed(
+    RadioClient* client,
+    RADIO_IND code,
+    const GBinderReader* args,
+    gpointer user_data)
+{
+    BinderRadioObject* self = THIS(user_data);
+    RADIO_STATE radio_state = RADIO_STATE_UNAVAILABLE;
+    GBinderReader reader;
+
+    /* radioStateChanged(RadioIndicationType, RadioState radioState); */
+   gbinder_reader_copy(&reader, args);
+   if (!gbinder_reader_read_int32(&reader, (gint32*) &radio_state)) {
```

Comment:
It sort of is, because `radio_types.h` has `G_STATIC_ASSERT(sizeof(RADIO_STATE) == 4);`. Well, it may not be 100% portable but when the assumption doesn't hold true, compilation will fail and I'm fine with that.

Reply:
Ahh, true, I missed that. Nevermind then, that is good enough for me.